pkg/controller/directimagemigration/imagestream.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2020 Red Hat Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package directimagemigration
+
+import (
+	"context"
+	//  liberr "github.com/konveyor/controller/pkg/error"
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	imagev1 "github.com/openshift/api/image/v1"
+	kapi "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (t *Task) getDirectImageStreamMigrations() ([]migapi.DirectImageStreamMigration, error) {
+	ismList := migapi.DirectImageStreamMigrationList{}
+	err := t.Client.List(
+		context.TODO(),
+		k8sclient.MatchingLabels(t.Owner.GetCorrelationLabels()),
+		&ismList)
+	if err != nil {
+		return nil, err
+	}
+	return ismList.Items, nil
+}
+
+func (t *Task) createDirectImageStreamMigrations() error {
+	// Get client for source
+	srcClient, err := t.getSourceClient()
+	if err != nil {
+		return err
+	}
+	// Get list namespaces to iterate over
+	for srcNsName, destNsName := range t.Owner.GetNamespaceMapping() {
+		isList := imagev1.ImageStreamList{}
+		err := srcClient.List(
+			context.TODO(),
+			k8sclient.InNamespace(srcNsName),
+			&isList)
+		if err != nil {
+			return err
+		}
+		for _, is := range isList.Items {
+			ismList := migapi.DirectImageStreamMigrationList{}
+			err := t.Client.List(
+				context.TODO(),
+				k8sclient.MatchingLabels(t.Owner.DirectImageStreamMigrationLabels(is)),
+				&ismList)
+			if err != nil {
+				return err
+			}
+			if len(ismList.Items) > 0 {
+				continue
+			}
+			imageStreamMigration := t.buildDirectImageStreamMigration(is, destNsName)
+			err = t.Client.Create(context.TODO(), &imageStreamMigration)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+func (t *Task) buildDirectImageStreamMigration(is imagev1.ImageStream, destNsName string) migapi.DirectImageStreamMigration {
+	labels := t.Owner.DirectImageStreamMigrationLabels(is)
+	imageStreamMigration := migapi.DirectImageStreamMigration{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:       labels,
+			GenerateName: t.Owner.GetName() + "-",
```

Comment:
Is there a reason why we need to use the generate name here and not use the same names as the imagestream?

I think we will have a more deterministic name with imagestream name and the list call here https://github.com/konveyor/mig-controller/pull/710/commits/18c8cb613353eb9db1024ab54811de43a805787c?file-filters%5B%5D=.go#diff-1ae7521ab52a30a2c3580aef36d5e819a554960ce543888a1d96548af27e0adfR50 can be replaced with a get call. (More performant as well I think not by much though)

Reply:
it would also be easier to debug, seeing the same imagestream name

Reply:
I was talking to @dymurray about this before I did this. There are two problems. First, we'd have to combine name and namespace into the name, which could run into length issues. Second, it would make it impossible to create a second CR -- say you did another migration to another cluster after the first. generateName eliminates these problems, because it isolates the CR to this particular migration. We can't hit name clashes because an already-finished (or even a currently-running) CR already exists with this name.